### ISSUE TITLE
Cosmos: CUD batching

### ISSUE THREAD

**Issue by AndriySvyryd** [dotnet/efcore#17308]
See https://github.com/Azure/azure-cosmos-dotnet-v3/blob/635cb3696490577b9fa115ec1e6f400c77c7b872/Microsoft.Azure.Cosmos.Samples/Usage/Batch/Program.cs
https://docs.microsoft.com/en-us/azure/cosmos-db/transactional-batch

Related: #36599

**Comment by vyarymovych** [comment]
Are there any updates?
Is it possible to persist a couple of cosmos documents in one transaction using EF Core Cosmos provider?

**Comment by roji** [comment]
@vyarymovych this issue is in the backlog, which means it's not currently in the plan for EF Core 7.0 - please vote (:+1:) for it on the top comment, that helps us prioritize work.

**Comment by PaulOst** [comment]
My feedback is that such a basic database feature, transactions, should not be in a backlog but planned for EF Core 7.0. Please reconsider. We can't use EF Core 7.0 on our projects without this capability.

**Comment by shanerogers** [comment]
We need this feature as well. The only way forward is to do state changes in our application with the CosmosClient with batches. Possibly use EF core to write queries, so we can leverage LINQ.

**Comment by rafalxyz** [comment]
Could this be prioritized please? The Cosmos provider is useless without such a basic feature. What's the point of using EF over Cosmos SDK without this feature?

**Comment by roji** [comment]
@rafalxyz there are many reasons to use the EF provider - change tracking, optimistic concurrency, etc. For now this is in the backlog, as it gets more votes we'll consider it when planning for new features.

**Comment by rafalxyz** [comment]
@roji I'm sorry for that comment. I've just used the EF provider in the latest commercial project and encountered so many bugs and missing features. Change tracking didn't work as expected either. I'll create an issue with everything that is missing or can be improved.

**Comment by PhilG0lding** [comment]
Can see this was a stretch goal for 9.0. I'd also like this to be prioritised too, as it would greatly simplify implementing an outbox pattern, as described here:
https://learn.microsoft.com/en-us/azure/architecture/databases/guide/transactional-outbox-cosmos

**Comment by JoasE** [comment]
The original post links to both TransactionalBatch and Bulk for CUD batching. As you likely know, Bulk and Batch are 2 very different things. From the thread I gather that most people (myself included) want TransactionalBatch support, not Bulk. 

As Bulk is designed for processing large volumes of data but increases latency I am not sure if enabling it (by default) in EF Core Cosmos DB provider would be wise. Though an option to enable it could be implemented, I think it should be a separate issue / feature request.

I would like to add that the provider feels incomplete without TransactionalBatch. Especially since other providers always use a transaction for DbContext commits, and the unit of work pattern usually implies or advocates for atomicity.

One challenge here is the fact that Cosmos DB does not support TransactionalBatch'es across partition keys. A simple solution for this might be listing this on the [Limitations page](https://learn.microsoft.com/en-us/ef/core/providers/cosmos/limitations). However, a (opt-in) configuration to throw on cross partition DbContext commits might be a nice addition to this feature for those looking to ensure atomicity.

**Comment by roji** [comment]
> I would like to add that the provider feels incomplete without TransactionalBatch. Especially since other providers always use a transaction for DbContext commits, and the unit of work pattern usually implies or advocates for atomicity.

I understand, but different providers support different things, and the fact that relational providers use transactions doesn't mean that a non-relational provider is incomplete if it doesn't. ACID guarantees and atomicity specifically are generally less of a central tenet of NoSQL/document database than it is in relational databases.

But of course, batching - and atomicity - in SaveChangesAsync are indeed important features that we intend to implement; it's mostly a matter of competing priorities and to what extent users tell us that this is something they need.

**Comment by PhilG0lding** [comment]
> > I would like to add that the provider feels incomplete without TransactionalBatch. Especially since other providers always use a transaction for DbContext commits, and the unit of work pattern usually implies or advocates for atomicity.
> 
> I understand, but different providers support different things, and the fact that relational providers use transactions doesn't mean that a non-relational provider is incomplete if it doesn't. ACID guarantees and atomicity specifically are generally less of a central tenet of NoSQL/document database than it is in relational databases.
> 
> But of course, batching - and atomicity - in SaveChangesAsync are indeed important features that we intend to implement; it's mostly a matter of competing priorities and to what extent users tell us that this is something they need.

Is there a formal place for logging this as a feature request or "upvoting"? Or is it just through these GitHub issues that you gain this feedback?

Currently as it stands - we have an application in the production environment using the Cosmos DB provider; It does a great job at abstracting the storage engine under the hood, however, as it stands, I wouldn't recommend other teams using it for applications in a Domain-Driven Design setting. This is purely due to the extra complexities we've had to overcome around aggregates and ACID principles.

**Comment by roji** [comment]
> Is there a formal place for logging this as a feature request or "upvoting"?

This Github issue is that formal place.

**Comment by JoasE** [comment]
I would like to pick this up (and have already done some work on it, not a problem if it needs to be thrown away it was fun to work on) as it seemed like a relatively small and contained change. My plan was to simply group changed entries by collection id and partition key and execute those groups as transactional batches. See my [draft PR](https://github.com/dotnet/efcore/pull/36590) for more details (still have some cleaning up and improving todo). However, I've run into conflicts with #19944. As far as I am aware, pre- and post-triggers are not supported in a transactional batch. It seems like we have to choose between them. Feedback from the team would be greatly appreciated. :)

**Comment by roji** [comment]
@JoasE am thinking out loud here... On relational/SQL databases, SaveChangesAsync guarantees that the *entire* batch is applied atomically. If I understand correctly, Cosmos only allows transactionality for operations within the same logical partition key ([docs](https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/transactional-batch?tabs=dotnet)). So with your proposed design, when an EF user executes SaveChangesAsync, some parts of it may execute in some transactions, others in others (because they affect different partitions).

In other words, Cosmos does not support full atomicity across partitions; this makes sense for a sharded/distributed-first database like Cosmos, but is a key difference from most relational databases (which aren't distributed and therefore can easily be transactional over their entire dataset). This, by the way, is exactly the kind of database difference I meant [in our previous discussion above](https://github.com/dotnet/efcore/issues/17308#issuecomment-3156044858) - one should not approach Cosmos (or other non-relational databases) with expectations that come from relational databases, and consider either Cosmos or its EF provider to be lacking if those expectations aren't met.

More to the point... I wonder if it makes sense for SaveChangesAsync to apply transactionality where it can, and what exactly the user gets out of that behavior. No full transactionality is guaranteed like in relational databases (because of the partitioning limitation), so you get something like "we'll use transactional batches where we can, best effort". Or in other words, if there's a failure in the middle of SaveChangesAsync, the user is left in a partially-committed state just as if no transactional batches were used at all. So I'm not sure what the benefits of this would be.

An alternative design would allow users to explicitly specify that they want transactional batches for SaveChangesAsync; when this is on, we could fail SaveChangesAsync if it can't be done atomically (i.e. if multiple partitions are modified). The user would then be responsible for structuring their SaveChangesAsync batches in a per-partition basis - this mirrors the way Cosmos actually works (and how users interact with transactional batches outside of EF).

@AndriySvyryd what do you think about all of this?

**Comment by roji** [comment]
BTW orthogonal to the above, we need to understand whether transactional batches provide a **performance** benefit regardless of atomicity. If they do (since less roundtrips are done), then I agree we should absolutely use them whenever we can, and then your proposed design of always grouping by container/partition ID and using batches makes sense (just not for atomicity - for performance). 

Finally, I don't think triggers are a big problem here - trigger usage is generally quite rare, and if a trigger is defined on a container, we could take that container out of the transactional batch. If we implement the atomic-only SaveChangesAsync opt-in as I proposed above, we'd error as these two features would be incompatible.

**Comment by JoasE** [comment]
@roji Thanks for the feedback!
I failed to mention that my current changes include an opt in to throw on cross partition/collection SaveChanges, which I briefly talked about in our previous discussion. It's a bit different but I think it's at-least in the same direction you seem to be thinking with explicitly specifying the use of transactional batches.

As to performance, there is definitely less overhead when using transactional batch (since less roundtrips are done), but memory usage on the client side might increase as more entities are stored in json in memory at the same time with a batch. I think this is a worthy tradeoff. I could not find any other official documentation on performance improvements when using transactional batches as oppose to individual requests, but Chat-GPT seems to suggest there might even be less RU consumption in some cases. (Not that it is always correct :p)

Regarding atomicity we also have another challenge in that cosmos db only supports up to 100 operations in a single batch.

Based on your feedback, my next plan would be to:
 - Find all entities with triggers defined
 - If the ThrowOnNonAtomicSaveChanges (or another name) is enabled and there are 2 or more entities and at least one of them has a trigger, throw an InvalidOperationException saying multiple entities with triggers can not be atomic
 - Create batches for all entities without triggers per 100 per collection per partition
 - If there is more than 1 batch and the ThrowOnNonAtomicSaveChanges is enabled, throw an InvalidOperationException saying cross collection/partition and more than 100 entities can not be atomic.
 - Execute the batches
 - save all entities with triggers with separate calls

I will wait on the feedback from @AndriySvyryd before continuing working on this. I might not make the 10.0 deadline (or feature cutoff) as I have limited time available.

I would like to note that having support for triggers does make the solution more complex due to having to maintain both the transactional batch and the separate calls code. Those will look a lot like each other but use different types and methods. And since you mention trigger usage is generally quite rare, what is the teams stance on these types of issues? al-tho I assume that the added complexity is negligible compared to the benefit of supporting an additional feature (even if it is used rarely).

**Comment by roji** [comment]
> I could not find any other official documentation on performance improvements when using transactional batches as oppose to individual requests

Is this something you can confirm with a quick benchmark? Basically just a super quick BenchmarkDotNet that compares regular insertion of e.g. 10 documents with and without a transactional batch? Ideally this would be done directly with the Cosmos SDK (no EF involved). I can also reach out to some Cosmos people internally to check, but it would be good to have some confirmed data first.

> I would like to note that having support for triggers does make the solution more complex due to having to maintain both the transactional batch and the separate calls code.

Don't you have to do that anyway because of the different partitionins (and the max 100) problems? In other words, I'd expect to have a relatively simple algorithm that takes all input changes to SaveChangesAsync and groups them in batches, taking into account:
1. Container/partition key
2. The 100 limit
3. Triggers (each change that touches a container that has a trigger defined on it would get its own single-item "batch", which we'd execute without an actual transactional batch)

Once you have the list of batches, if the user has opted into ThrowOnNonAtomicSaveChanges, throw if there's more than one batch. Otherwise, just execute them. With this logic, triggers don't seem to add any significant complexity, no?

> I will wait on the feedback from @AndriySvyryd before continuing working on this. I might not make the 10.0 deadline (or feature cutoff) as I have limited time available.

Note that the ship has sailed for new feature work in 10 - we just branched for rc.1... So this will in any case need to wait for EF 11. No rush.

**Comment by JoasE** [comment]
> Is this something you can confirm with a quick benchmark? Basically just a super quick BenchmarkDotNet that compares regular insertion of e.g. 10 documents with and without a transactional batch? Ideally this would be done directly with the Cosmos SDK (no EF involved). I can also reach out to some Cosmos people internally to check, but it would be good to have some confirmed data first.

I can definitely try. Could you clarify what you’d like the benchmark to measure: RU consumption, general performance or performance excluding latency/overhead? If RU/s don’t differ, I’m not sure we’ll be able to attribute performance differences solely to request overhead versus something else (e.g. shared indexing). But if the goal is just to see whether simply using transactional batch has a meaningful impact, a simple benchmark should be sufficient.

> Don't you have to do that anyway because of the different partitionins (and the max 100) problems?

I don't think so, the cosmos sdk uses different CosmosClient methods and RequestOptions classes for single Create/Update/Delete calls as oppose transactional batch. However, maybe I'm over exaggerating a bit on the 'complexity' that will bring. It will be some 'code duplication' against the different options classes at-least and some refactoring to be able to share code between the single calls path and the batch path. I'm fine with doing it that way and seeing what the end result will be. If I hit blockers, would it be OK to share early for guidance?

**Comment by roji** [comment]
> Could you clarify what you’d like the benchmark to measure: RU consumption, general performance or performance excluding latency/overhead?

All of these are relevant, I think - if this really does reduce roundtrips, then this should pretty dramatically reduce the end-to-end overall operation time (potentially one roundtrip instead of X). And of course, RU savings are great too.

I'm asking since I'm not quite sure how exactly the transaction batching being discussed here compares to bulk operations [as described here](https://devblogs.microsoft.com/cosmosdb/introducing-bulk-support-in-the-net-sdk/).

> I don't think so, the cosmos sdk uses different CosmosClient methods and RequestOptions classes for single Create/Update/Delete calls as oppose transactional batch.

Of course; my point was rather that you have to worry about batch splitting in any case (because of partitioning, 100 max limit), so it seems like adding the additional splitting reason (trigger) shouldn't be a meaningful increase in complexity.

**Comment by AndriySvyryd** [comment]
Unless the benchmarks prove otherwise, it seems there's no significant downside to doing best-effort batching as described above, so I am ok with always doing it.
As per the atomicity exception - we can issue a configurable warning whenever one of the above conditions prevents an atomic `SaveChangesAsync` execution. It will just be logged by default, but if the users configure the warnings to throw by default or set these specific warnings to throw then an exception would be thrown.
We can have different message strings for each case, but use the same EventId for ease of use

**Comment by roji** [comment]
@AndriySvyryd one thought there, is that users may want to sometimes perform transactional SaveChangesAsync and sometimes non-transactional, in the same application. The problem with suppressible warnings is that it's a global setting that can't be changed on a per-invocation basis.

This is all a bit similar to the [AutoTransactionBehavior](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.autotransactionbehavior?view=efcore-9.0) setting that we allow configuring on context.Database. We could consider repurposing that for this Cosmos usage (it's not like that enum means anything else for Cosmos); it isn't great, since the meaning is different in Cosmos, where we need to allow the user to configure whether to throw if the SaveChangesAsync can't be performed in a single transaction, and maybe to determine whether transactions should be used or not even when they don't make the entire SaveChangesAsync atomic (still TBD).

Regardless of whether we repurpose AutoTransactionBehavior though, my main point is that I think we should allow this to be manageable on a per-invocation basis rather than as a global setting...

**Comment by AndriySvyryd** [comment]
@roji Sure, we can map it as "Don't throw", "Throw", "No transactions"

**Comment by roji** [comment]
Yeah, the naming isn't perfect (WhenNeeded -> use transactional batches where you can, Always -> Throw if you can't do a single transactional batch, Never -> never use transactional batches), but I agree it's probably OK.

I'd still want to see a tangible performance benefit to transactional batches, otherwise using them sometimes/when it works doesn't sound like a great idea.

**Comment by JoasE** [comment]
@roji @AndriySvyryd 
I ran the following console apps on my local machine against a real single region cosmos db hosted in azure: https://github.com/JoasE/CosmosBenchmark

BenchmarkDotNet tests:

10 items:
| Method                          | Mean      | Error    | StdDev   |
|-------------------------------- |----------:|---------:|---------:|
| SingleSequentialInsertsStream   | 302.13 ms | 4.707 ms | 3.675 ms |
| SingleParallelInsertsStream     |  65.18 ms | 1.303 ms | 1.827 ms |
| TransactionalBatchInsertsStream |  36.03 ms | 0.705 ms | 1.234 ms |

100 items:
| Method                          | Mean        | Error     | StdDev    |
|-------------------------------- |------------:|----------:|----------:|
| SingleSequentialInsertsStream   | 2,816.41 ms | 32.555 ms | 28.859 ms |
| SingleParallelInsertsStream     |   124.37 ms |  2.453 ms |  5.121 ms |
| TransactionalBatchInsertsStream |    85.62 ms |  1.616 ms |  1.349 ms |

RU tests (ran multiple times with same output):
10 items:
RU single 62,9
RU batch 62,86

100 items:
RU single 629,0000000000001
RU batch 628,57

I think we can conclude that TransactionalBatch is more performant in RU usage as well as time to execute, even when running single inserts in parallel. I don't think it will be necessary to run this from an azure vm or webapp in the same aswell, though the difference in speed might be less significant due to less network latency (azure backbone).

Let me know if you see issues with the benchmarks or have any other questions

**Comment by JoasE** [comment]
> This is all a bit similar to the [AutoTransactionBehavior](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.autotransactionbehavior?view=efcore-9.0) setting that we allow configuring on context.Database. We could consider repurposing that for this Cosmos usage (it's not like that enum means anyhting else for Cosmos); it isn't great, since the meaning is different in Cosmos, where we need to allow the user to configure whether to throw if the SaveChangesAsync can't be performed in a single transaction, and maybe to determine whether transactions should be used or not even when they don't make the entire SaveChangesAsync atomic (still TBD).

I have some reservations about reusing enums this way, but I trust your judgment. Maybe offtopic, but have you guys considered moving more relational specific configurations like AutoTransactionBehavior to Microsoft.EntityFrameworkCore.Relational (maybe defining extension methods on DatabaseFacade)? Could we use something along the lines of that for the cosmos transaction behaviour?

Based on the current discussion my plan for this feature would be to (psuedo code):
```csharp
var entitiesWithoutTriggers = allEntities.Where(e => !e.HasTrigger).ToList();
var entitiesWithTriggers = allEntities.Where(e => e.HasTrigger).ToList();

if (transactionBehavior == TransactionBehavior.Throw &&
    entitiesWithTriggers.Count >= 1 && allEntities.Count >= 2)
{
    throw new InvalidOperationException("Multiple entities with triggers cannot be saved atomically.");
}

foreach (var entity in entitiesWithTriggers)
{
    SaveEntity(entity);
}

if (transactionBehavior == TransactionBehavior.NoTransactions)
{
    foreach (var entity in entitiesWithoutTriggers)
    {
        SaveEntity(entity);
    }
    return;
}

var batches = CreateBatches(entitiesWithoutTriggers); // group by collection, partition and max 100

if (transactionBehavior == TransactionBehavior.Throw && batches.Count > 1)
{
    throw new InvalidOperationException("Cross collection/partition or >100 entities cannot be saved atomically.");
}

foreach (var batch in batches)
{
    ExecuteBatch(batch);
}
```

**Comment by JoasE** [comment]
Maybe a minor detail, but this line:
> if (transactionBehavior == TransactionBehavior.NoTransactions)

Could also be:
`if (transactionBehavior == TransactionBehavior.NoTransactions || entitiesWithoutTriggers.Count == 1)`
As there should be no difference on atomicity.
However, this wouldn't really fit the naming of  `AutoTransactionBehavior.Always`, but would fit `TransactionBehavior.Throw` or `CosmosTransactionBehavior.Throw`

**Comment by JoasE** [comment]
@roji @AndriySvyryd I’ve opened a draft PR that’s almost ready for review: https://github.com/dotnet/efcore/pull/36693. I’d appreciate some guidance from the team regarding logging, specifically at [CosmosClientWrapper.cs:660](https://github.com/dotnet/efcore/pull/36693/files#diff-28e91c7cb261ce53c36e0e2d3f1ba871561ad484a0452ea5044c51b6daa9c63aR660).

The batch executes as a single operation but contains multiple entries.

Should the batch execution result in one log entry (covering the whole batch), or multiple log entries (one per item in the batch)?

If a single log entry is preferred, what format should it take (e.g., how should I represent the item IDs and partition keys)?

If multiple log entries are preferred, should I define something like ExecutedBatchCreateItem? And in that case, how should I handle fields such as ActivityId, RequestCharge, or ElapsedTime, which apply to the batch rather than individual items?

**Comment by JoasE** [comment]
I also have a question regarding the suppression of DbUpdateConcurrencyException, see [CosmosDatabaseWrapper.cs:155](https://github.com/dotnet/efcore/pull/36693/files#diff-139c85f555914040829e8f8ee570a78fb158412817e160e9f46d7e7406510c3eR155) 
If I understand correctly, this means that we will have to recreate and retry the batch without the entries causing the concurrency issues, but I would like to double check with the team. Retrying can fail due to concurrency again, should we keep retrying until there are are no entries left in the batch, or fail sooner?

**Comment by AndriySvyryd** [comment]
> I have some reservations about reusing enums this way, but I trust your judgment. Maybe offtopic, but have you guys considered moving more relational specific configurations like AutoTransactionBehavior to Microsoft.EntityFrameworkCore.Relational (maybe defining extension methods on DatabaseFacade)? Could we use something along the lines of that for the cosmos transaction behaviour?

I think it's a bit late for this sort of change, even though `AutoTransactionBehavior` didn't do anything for non-relational.

> Could also be:
`if (transactionBehavior == TransactionBehavior.NoTransactions || entitiesWithoutTriggers.Count == 1)`

It's up to you, we don't need to be pedantic in this case.
```
if (transactionBehavior == TransactionBehavior.Throw && batches.Count > 1)
{
    throw new InvalidOperationException("Cross collection/partition or >100 entities cannot be saved atomically.");
}
```
Make sure that all these deterministic exceptions are thrown before anything is actually sent to the database

**Comment by AndriySvyryd** [comment]
> If I understand correctly, this means that we will have to recreate and retry the batch without the entries causing the concurrency issues, but I would like to double check with the team. Retrying can fail due to concurrency again, should we keep retrying until there are no entries left in the batch, or fail sooner?

No, if the exception is suppressed just carry on as if no exception was thrown. It's up to the user the ensure that the interceptor suppresses the exception only in the expected cases.

**Comment by JoasE** [comment]
@AndriySvyryd @roji 

Wondering if this might have been an oversight on our part or if you are ok with leaving the management of the batch size up to the user. The below succeeds with 10.0 or 11.0 with AutoTransactionBehavior.Never, but fails with 11.0 AutoTransactionBehavior.WhenNeeded

```
[ConditionalFact]
public virtual async Task SaveChanges_three_1mb_entries_succeeds()
{
    var contextFactory = await InitializeAsync<TransactionalBatchContext>();

    using var context = contextFactory.CreateContext();

    context.Customers.Add(new Customer { Id = "1", Name = new string('x', 1_000_000), PartitionKey = "1" });
    context.Customers.Add(new Customer { Id = "2", Name = new string('x', 1_000_000), PartitionKey = "1" });
    context.Customers.Add(new Customer { Id = "3", Name = new string('x', 1_000_000), PartitionKey = "1" });

    await context.SaveChangesAsync(); // Fails: DbUpdateException: CosmosException: Response status code does not indicate success: RequestEntityTooLarge 

    using var assertContext = contextFactory.CreateContext();
    var customersCount = await assertContext.Customers.CountAsync();
    Assert.Equal(3, customersCount);
}
```

We could get the current batch entry size as we are serializing to a stream already, and add a safe and/or calculated margin for request metadata. Then EF Core could automatically split the above into 2 batches, and only throw with AutoTransactionBehavior.Always

If so, should I create a new issue?

Also see: Azure/azure-cosmos-dotnet-v3#5406

**Comment by AndriySvyryd** [comment]
> If so, should I create a new issue?

Please do
